trino/integration_test.go

```diff
@@ -1008,82 +1008,123 @@ func TestIntegrationUnsupportedHeader(t *testing.T) {
+	}{
+		{
+			name:           "Context Cancellation",
+			timeout:        0,
+			manualCancel:   true,
+			expectedErrMsg: "canceled",
+		},
+		{
+			name:           "Context Deadline Exceeded",
+			timeout:        3 * time.Second,
+			manualCancel:   false,
+			expectedErrMsg: "context deadline exceeded",
+		},
 	}
-	dsn := *integrationServerFlag
-	dsn += "?catalog=tpch&schema=sf100&source=cancel-test&custom_client=uncompressed"
-	db := integrationOpen(t, dsn)
-	defer db.Close()
 
-	ctx, cancel := context.WithCancel(context.Background())
-	errCh := make(chan error, 3)
-	done := make(chan struct{})
-	longQuery := "SELECT COUNT(*) FROM lineitem"
-	go func() {
-		// query will complete in ~7s unless cancelled
-		rows, err := db.QueryContext(ctx, longQuery)
-		if err != nil {
-			errCh <- err
-			return
-		}
-		rows.Next()
-		if err = rows.Err(); err != nil {
-			errCh <- err
-			return
-		}
-		close(done)
-	}()
-
-	// poll system.runtime.queries and wait for query to start working
-	var queryID string
-	pollCtx, pollCancel := context.WithTimeout(context.Background(), 1*time.Second)
-	defer pollCancel()
-	for {
-		row := db.QueryRowContext(pollCtx, "SELECT query_id FROM system.runtime.queries WHERE state = 'RUNNING' AND source = 'cancel-test' AND query = ?", longQuery)
-		err := row.Scan(&queryID)
-		if err == nil {
-			break
-		}
-		if err != sql.ErrNoRows {
-			t.Fatal("failed to read query id", err)
-		}
-		if err = contextSleep(pollCtx, 100*time.Millisecond); err != nil {
-			t.Fatal("query did not start in 1 second")
-		}
-	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := RegisterCustomClient("uncompressed", &http.Client{Transport: &http.Transport{DisableCompression: true}})
+			if err != nil {
+				t.Fatal(err)
+			}
 
-	cancel()
+			dsn := *integrationServerFlag
+			dsn += "?catalog=tpch&schema=sf100&source=cancel-test&custom_client=uncompressed"
+			db := integrationOpen(t, dsn)
+			defer db.Close()
 
-	select {
-	case <-done:
-		t.Fatal("unexpected query with cancelled context succeeded")
-		break
-	case err = <-errCh:
-		if !strings.Contains(err.Error(), "canceled") {
-			t.Fatal("expected err to be canceled but got:", err)
-		}
-	}
+			var ctx context.Context
+			var cancel context.CancelFunc
 
-	// poll system.runtime.queries and wait for query to be cancelled
-	pollCtx, pollCancel = context.WithTimeout(context.Background(), 1*time.Second)
-	defer pollCancel()
-	for {
-		row := db.QueryRowContext(pollCtx, "SELECT state, error_code FROM system.runtime.queries WHERE query_id = ?", queryID)
-		var state string
-		var code *string
-		err := row.Scan(&state, &code)
-		if err != nil {
-			t.Fatal("failed to read query id", err)
-		}
-		if state == "FAILED" && code != nil && *code == "USER_CANCELED" {
-			break
-		}
-		if err = contextSleep(pollCtx, 100*time.Millisecond); err != nil {
-			t.Fatal("query was not cancelled in 1 second; state, code, err are:", state, code, err)
-		}
+			if tt.manualCancel {
```

Comment:
I don't think we need a separate field like this, it would be enough to check if timeout is set (is non-zero).